redpanda/validators/throughput_tiers.go

```diff
@@ -0,0 +1,213 @@
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// ThroughputTierValidator is a custom validator that ensures that the throughput tier is valid for the given cloud provider, region, and zones
+type ThroughputTierValidator struct{}
+
+// Description provides a description of the validator
+func (ThroughputTierValidator) Description(_ context.Context) string {
+	return "ensures that the throughput tier is valid for the given cloud provider, region, and zones"
+}
+
+// MarkdownDescription provides a description of the validator in markdown format
+func (ThroughputTierValidator) MarkdownDescription(_ context.Context) string {
+	return "Ensures that the throughput tier is valid for the given cloud provider, region, and zones"
+}
+
+// ValidateString validates a string
+func (ThroughputTierValidator) ValidateString(ctx context.Context, req validator.StringRequest, resp *validator.StringResponse) {
+	if req.ConfigValue.IsUnknown() || req.ConfigValue.IsNull() {
+		return
+	}
+
+	var zones types.List
+	var cloudProvider, region types.String
+	resp.Diagnostics.Append(req.Config.GetAttribute(ctx, path.Root("cloud_provider"), &cloudProvider)...)
+	resp.Diagnostics.Append(req.Config.GetAttribute(ctx, path.Root("region"), &region)...)
+	resp.Diagnostics.Append(req.Config.GetAttribute(ctx, path.Root("zones"), &zones)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	throughputTier := req.ConfigValue.ValueString()
+
+	// Validate the throughput tier
+	validTiers := make(map[string]bool)
+	invalidZones := make([]string, 0)
+
+	for _, zoneElement := range zones.Elements() {
+		zone := zoneElement.(types.String).ValueString()
+		if tiers, ok := validThroughputTiers[cloudProvider.ValueString()][region.ValueString()][zone]; ok {
+			for _, tier := range tiers {
+				validTiers[tier] = true
+			}
+		} else {
+			invalidZones = append(invalidZones, zone)
+		}
+	}
+
+	if len(invalidZones) > 0 {
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			"Invalid Zones",
+			fmt.Sprintf("The following zones are not valid for the given cloud provider (%s) and region (%s): %v",
+				cloudProvider.ValueString(), region.ValueString(), invalidZones),
+		)
+		return
+	}
+
+	if _, isValid := validTiers[throughputTier]; !isValid {
+		resp.Diagnostics.AddAttributeError(
+			req.Path,
+			"Invalid Throughput Tier",
+			fmt.Sprintf("The throughput tier %s is not valid for the given cloud provider (%s), region (%s), and zones. Valid tiers are: %v",
+				throughputTier, cloudProvider.ValueString(), region.ValueString(), getKeys(validTiers)),
+		)
+	}
+}
+
+// Helper function to get keys from a map
+func getKeys(m map[string]bool) []string {
+	keys := make([]string, 0, len(m))
+	for k := range m {
+		keys = append(keys, k)
+	}
+	return keys
+}
+
+var validThroughputTiers = map[string]map[string]map[string][]string{
```

Comment:
Is there a way to get this list through the controlplane API?

This list is quite big and might change often and we might end up blocking customers from creating a cluster using the Provider if we don't have this updated. Also, there is no communication/process in place to update this list if Cloud adds/removes a new Tier.

Reply:
My understanding is that it hasn't been added yet as an endpoint so I am awaiting it.

Reply:
It's available now in https://buf.build/redpandadata/cloud/docs/main:redpanda.api.controlplane.v1beta2#redpanda.api.controlplane.v1beta2.ThroughputTierService

You'll have to bump the version to use it.

Reply:
Yes, it was just added very recently. Asked them about it in the public api call